### DISCUSSION TITLE
how to deal with exceptions when writing json?

### DISCUSSION THREAD

**Discussion by scalpel4k** [stephenberry/glaze#674] [General]
Hi,

I see that the write operations are all noexcept. Is there a way to abort writing when some conditions fail using the custom serializers?

thx Michi

**Comment by stephenberry** [comment]
Can you use [std::longjmp](https://en.cppreference.com/w/cpp/utility/program/longjmp)?

Currently Glaze does not provide a mechanism for this. However, conditional exception logic within the glaze compile time options is probably a good design choice, so I will be considering how to accomplish this in the future.

**Reply by scalpel4k** [reply]
thx for the reply. Actually, what I'd like to avoid is that the caller of the serializing has to do some "unfamiliar" error handling.

An obvious choice for me would be for the write function to either throw an exception or return a std::optional (or when using the external buffer approach some return code instead of void). Naively, I would say that the write function could be declared noexcept(false) just like the read function ....

**Reply by stephenberry** [reply]
I think returning an error code is good for errors that do not corrupt program memory (clarification: errors that would not lead to corrupting memory). For corrupting errors I think exceptions should be supported or perhaps std::longjmp is better to maintain noexcept. I don't want to take the same exact approach as reading, because errors in writing should be much less common and writing all the short-circuiting logic is both more complex and slower at runtime.

We can also provide a compile time option `noexcept(!Opts.allow_exceptions)`, but I would prefer error handling to be the same across use cases.

**Reply by scalpel4k** [reply]
I fully understand your reasoning. My use case would be to serialize JSON data that must adhere to a protocol spec/schema. So write errors in that sense are more like logic errors not type errors. And I think these can be just as common ...

Maybe another solution would be to add `try_write` functions, so the noexcept mechanism could be left untouched.

**Reply by stephenberry** [reply]
Do you use JSON schema? In a lot of cases we can validate the C++ structures in Glaze at compile time and ensure that the messages generated would conform to the protocol spec/schema. This would mean there would be no runtime validation cost. I've been considering adding this to glaze, as we are part of the way there already. Is this something that would be useful for you, or do you need runtime validation for writing?

**Reply by scalpel4k** [reply]
now that you're asking I would not complain about compile-time validation, however most use cases that come to my mind would require runtime validation

**Reply by stephenberry** [reply]
Cool, I'll make an issue concerning writing validation. And, as much as possible we can try to synthesize the compile time and runtime validation.

**Reply by scalpel4k** [reply]
thanks a lot for the discussion
